File: app/src/main/scala/com/github/pflooky/datagen/core/util/TrialUtil.scala
package com.github.pflooky.datagen.core.util

import com.github.pflooky.datacaterer.api.model.Constants.API_KEY
import com.github.pflooky.datagen.core.model.Constants.DATA_CATERER_SITE_PRICING
import org.apache.log4j.Logger
import org.joda.time.{DateTime, Days}

import java.util.Base64
import javax.crypto.Cipher
import javax.crypto.spec.{IvParameterSpec, SecretKeySpec}
import scala.util.{Failure, Success, Try}

object TrialUtil {

  private val LOGGER = Logger.getLogger(getClass.getName)
  private val pattern = "^.+?random ([0-9]{4}-[0-9]{2}-[0-9]{2}) text.*$".r
  private val numberOfTrialDays = 31

  def checkApiKey(key: String = ""): Boolean = {
    val apiKeyEnv = if (key.nonEmpty) key else System.getenv(API_KEY)
    if (apiKeyEnv == null) {
      LOGGER.error(s"$API_KEY is not defined as an environment variable. Please define to run Data Caterer")
      false
    } else {
      val decryptedApiKey = new KeyEncryption().decrypt(apiKeyEnv)
      decryptedApiKey match {
        case Some(pattern(date)) =>
          val parsedDate = DateTime.parse(date)
          val today = DateTime.now()
          val daysBetween = Days.daysBetween(DateTime.now, parsedDate.plusDays(numberOfTrialDays)).getDays
          val daysLeft = Math.max(daysBetween, 0)
          if (daysLeft == 0) {
            LOGGER.error(s"Your trial period has expired. Please consider upgrading to the paid plan to help support development $DATA_CATERER_SITE_PRICING")
          } else {
            LOGGER.info(s"Your trial has $daysLeft days left. Go nuts and try everything out!")
          }
          daysLeft > 0
        case _ =>
          LOGGER.error("Failed to verify API key")
          false
      }
    }
  }

  class KeyEncryption {
    private val key = "getBytesgetStringgetInst"
    private val initVector = "getInstanceInitt"
    private val iv = new IvParameterSpec(initVector.getBytes("UTF-8"))
    private val skeySpec = new SecretKeySpec(key.getBytes("UTF-8"), "AES")
    private val cipher = Cipher.getInstance("AES/CBC/PKCS5PADDING")

    def encrypt(string: String): String = {
      cipher.init(Cipher.ENCRYPT_MODE, skeySpec, iv)
      val encrypted = cipher.doFinal(string.getBytes)
      Base64.getEncoder.encodeToString(encrypted)
    }

    def decrypt(string: String): Option[String] = {
      cipher.init(Cipher.DECRYPT_MODE, skeySpec, iv)
      val tryDecrypt = Try(cipher.doFinal(Base64.getDecoder.decode(string)))
      tryDecrypt match {
        case Failure(_) =>
          LOGGER.error("Invalid API_KEY. Please reach out to Peter Flook on Slack for assistance")
          None
        case Success(value) =>
          Some(new String(value))
      }
    }
  }
}

File: trial/src/main/java/com/github/pflooky/datacatererutils/slack/TrialUtil.java
package com.github.pflooky.datacatererutils.slack;

import com.slack.api.app_backend.slash_commands.response.SlashCommandResponse;
import com.slack.api.bolt.context.builtin.SlashCommandContext;
import com.slack.api.bolt.request.builtin.SlashCommandRequest;
import org.joda.time.DateTime;
import org.joda.time.Days;

import javax.crypto.Cipher;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.StandardOpenOption;
import java.util.Base64;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

public class TrialUtil {

    private final Map<String, String> userKeys;
    private final Pattern pattern = Pattern.compile("^.+?random ([0-9]{4}-[0-9]{2}-[0-9]{2}) text.*$");
    private final Path userKeysFilePath = Paths.get("/home/ec2-user/user-keys.csv");
    private final int numberOfTrialDays = 31;
    private final KeyEncryption keyEncryption = new KeyEncryption();

    public TrialUtil() throws Exception {
        this.userKeys = initUserKeyMap();
    }

    public SlashCommandResponse generateApiKey(SlashCommandRequest req, SlashCommandContext ctx) {
        SlashCommandResponse slashCommandResponse = new SlashCommandResponse();
        slashCommandResponse.setResponseType("ephemeral");
        try {
            String apiKey;
            if (isExistingUser(ctx)) {
                apiKey = userKeys.get(ctx.getRequestUserId());
            } else {
                apiKey = generateApiKey();
                userKeys.put(ctx.getRequestUserId(), apiKey);
                String newFileEntry = ctx.getRequestUserId() + "," + apiKey + "\n";
                Files.write(userKeysFilePath, newFileEntry.getBytes(), StandardOpenOption.APPEND);
            }

            Optional<DateTime> optionalDateTime = getDate(keyEncryption.decrypt(apiKey));
            int daysLeft = optionalDateTime.map(dateTime -> {
                int daysBetween = Days.daysBetween(DateTime.now(), dateTime.plusDays(numberOfTrialDays)).getDays();
                return Math.max(daysBetween, 0);
            }).orElse(0);
            String message;
            if (daysLeft == 0) {
                message = "Your trial period has expired. Please consider upgrading to the paid plan to help support development https://data.catering/pricing/";
            } else {
                message = "*Your API key is:* `" + apiKey + "`\n*Number of days left for trial:* " + daysLeft + " days";
            }
            slashCommandResponse.setText(message);
        } catch (Exception ex) {
            slashCommandResponse.setText("Failed to generate new API key. Please contact Peter Flook in Slack for support.");
        }
        return slashCommandResponse;
    }

    public String generateApiKey() throws Exception {
        String today = new DateTime().toString("yyyy-MM-dd");
        String baseText = "some% random " + today + " text˚¬å";
        return keyEncryption.encrypt(baseText);
    }

    public boolean isExistingUser(SlashCommandContext ctx) {
        return userKeys.containsKey(ctx.getRequestUserId());
    }

    private Optional<DateTime> getDate(String key) {
        Matcher matcher = pattern.matcher(key);
        if (matcher.matches()) {
            String date = matcher.group(1);
            return Optional.of(DateTime.parse(date));
        } else {
            return Optional.empty();
        }
    }

    private Boolean checkApiKey(String key) throws Exception {
        String apiKeyEnv = key.length() > 0 ? key : System.getenv("API_KEY");
        if (apiKeyEnv == null) {
            return false;
        } else {
            String decryptedApiKey = keyEncryption.decrypt(apiKeyEnv);
            Matcher matcher = pattern.matcher(decryptedApiKey);
            if (matcher.matches()) {
                String date = matcher.group(1);
                DateTime parsedDate = DateTime.parse(date);
                DateTime today = DateTime.now();
                return parsedDate.plusDays(numberOfTrialDays).isBefore(today);
            } else {
                return false;
            }
        }
    }

    private Map<String, String> initUserKeyMap() {
        try {
            List<String> lines = Files.readAllLines(userKeysFilePath);
            return lines.stream().map(line -> {
                String[] spt = line.split(",");
                return Map.entry(spt[0], spt[1]);
            }).collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
        } catch (Exception ex) {
            return Collections.emptyMap();
        }
    }

    static class KeyEncryption {
        private final String key = "getBytesgetStringgetInst";
        private final String initVector = "getInstanceInitt";
        private final IvParameterSpec iv = new IvParameterSpec(initVector.getBytes(StandardCharsets.UTF_8));
        private final SecretKeySpec skeySpec = new SecretKeySpec(key.getBytes(StandardCharsets.UTF_8), "AES");
        private final Cipher cipher = Cipher.getInstance("AES/CBC/PKCS5PADDING");

        KeyEncryption() throws Exception {
        }

        String encrypt(String string) throws Exception {
            cipher.init(Cipher.ENCRYPT_MODE, skeySpec, iv);
            byte[] encrypted = cipher.doFinal(string.getBytes());
            return Base64.getEncoder().encodeToString(encrypted);
        }

        String decrypt(String string) throws Exception {
            cipher.init(Cipher.DECRYPT_MODE, skeySpec, iv);
            byte[] decrypted = cipher.doFinal(Base64.getDecoder().decode(string));
            return new String(decrypted);
        }
    }
}
